### ISSUE TITLE
Use refs for non-simple chained comparisons

### ISSUE THREAD

**Issue by STRd6** [DanielXMoore/Civet#50]
Use a ref in non-simple chained comparisons so not to repeat evaluations of complex expressions.

```civet
f1(x) < f2(x) < f3(x) < f4(x)
```

```
let ref
f1(x) < (ref = f2(x)) && ref < (ref = f3(x)) && ref < f4(x)
```

**Comment by edemaine** [comment]
Other settings where we should use refs (please add to list if you encounter any):

* `{ foo[x+y] }` → `{ [x+y]: foo[x+y] }` (#284)

**Comment by JayXon** [comment]
I see that this issue is tagged as efficiency, but I think this is not just efficiency, but a correctness issue, since an expression could have side effect, see the examples in #1341 and #1344

**Comment by JayXon** [comment]
it seems #1344 is still not fixed, can we keep this open or reopen #1344?
